### ISSUE TITLE
Vsphere datastore cluster support

### ISSUE THREAD

**Issue by hashibot** [terraform-providers/terraform-provider-vsphere#2]
_This issue was originally opened by @hugoboos as hashicorp/terraform#3721. It was migrated here as part of the [provider split](https://www.hashicorp.com/blog/upcoming-provider-changes-in-terraform-0-10/). The original body of the issue is below._

<hr>

Hi,

It looks like the Vsphere provider does not have support for datastore clusters.
When I provide a datastore cluster name as `datastore` under the `disk` section, Terraform reports back that it can't find the datastore.

With `govc` from https://github.com/vmware/govmomi I can list the the datastores with the command `govc ls`.

Thanks!

**Comment by hashibot** [comment]
_This comment was originally opened by @tkak as https://github.com/hashicorp/terraform/issues/3721#issuecomment-154825539. It was migrated here as part of the [provider split](https://www.hashicorp.com/blog/upcoming-provider-changes-in-terraform-0-10/). The original comment is below._

<hr>

Current version supports datastore cluster feature partly. If you don't mind, could you tell me your use cases more detail? (.tf file, datastore directory path, etc)

**Comment by hashibot** [comment]
_This comment was originally opened by @hugoboos as https://github.com/hashicorp/terraform/issues/3721#issuecomment-155849843. It was migrated here as part of the [provider split](https://www.hashicorp.com/blog/upcoming-provider-changes-in-terraform-0-10/). The original comment is below._

<hr>

We have a VMware cluster with resources pools, storage DRS. The whole enterprise thing.

What we want is to deploy virtual machines based on cluster names, resource pool names, datastore cluster names, etc. like below in the `main.tf` file. We do not want to specify the directory path but only the name of the object.

Regarding storage, Storage DRS should decide on which datastore the virtual machine ends up.

``` terraform
# main.tf

provider "vsphere" {
  vcenter_server = "vcenter"
  user           = "DOMAIN\account"
  password       = "password"
}

resource "vsphere_virtual_machine" "webserver01" {
  name   = "webserver01"
  vcpu   = 4
  memory = 2048

  datacenter    = "datacenter name"
  cluster       = "cluster name"
  resource_pool = "resource pool name"

  time_zone = "Europe/Amsterdam"

  network_interface {
    label = "network name"
  }

  disk {
    datastore = "datastore cluster name"
    size      = 40
  }
}
```

**Comment by hashibot** [comment]
_This comment was originally opened by @chrislovecnm as https://github.com/hashicorp/terraform/issues/3721#issuecomment-158691844. It was migrated here as part of the [provider split](https://www.hashicorp.com/blog/upcoming-provider-changes-in-terraform-0-10/). The original comment is below._

<hr>

@tkak what changes would need to take place? 

@mijit thoughts?

**Comment by hashibot** [comment]
_This comment was originally opened by @kc11381 as https://github.com/hashicorp/terraform/issues/3721#issuecomment-159290002. It was migrated here as part of the [provider split](https://www.hashicorp.com/blog/upcoming-provider-changes-in-terraform-0-10/). The original comment is below._

<hr>

Hi,

For this we need support at govmomi library itself. https://github.com/vmware/govmomi/blob/master/find/finder.go look at line number 255.
I think we should have something 
       if ref.Type == "Datastore"  || ref.Type == "StoragePod"{

I have done it for my use case and it works fine.

**Comment by hashibot** [comment]
_This comment was originally opened by @chrislovecnm as https://github.com/hashicorp/terraform/issues/3721#issuecomment-159450487. It was migrated here as part of the [provider split](https://www.hashicorp.com/blog/upcoming-provider-changes-in-terraform-0-10/). The original comment is below._

<hr>

@kc11381 you mind filing an issue with Govmomi and referencing it here?

**Comment by hashibot** [comment]
_This comment was originally opened by @kc11381 as https://github.com/hashicorp/terraform/issues/3721#issuecomment-159469280. It was migrated here as part of the [provider split](https://www.hashicorp.com/blog/upcoming-provider-changes-in-terraform-0-10/). The original comment is below._

<hr>

Hi Chris,

Here is the ticket:
https://github.com/vmware/govmomi/issues/386
Thank you.

**Comment by hashibot** [comment]
_This comment was originally opened by @hugoboos as https://github.com/hashicorp/terraform/issues/3721#issuecomment-160654153. It was migrated here as part of the [provider split](https://www.hashicorp.com/blog/upcoming-provider-changes-in-terraform-0-10/). The original comment is below._

<hr>

Cool.

If we need to test something, please let me know. I'm willing to help out.

**Comment by hashibot** [comment]
_This comment was originally opened by @daryn-kerr as https://github.com/hashicorp/terraform/issues/3721#issuecomment-182957174. It was migrated here as part of the [provider split](https://www.hashicorp.com/blog/upcoming-provider-changes-in-terraform-0-10/). The original comment is below._

<hr>

anyone had any success with this? would be great to get it working.

**Comment by hashibot** [comment]
_This comment was originally opened by @mixacha as https://github.com/hashicorp/terraform/issues/3721#issuecomment-183070392. It was migrated here as part of the [provider split](https://www.hashicorp.com/blog/upcoming-provider-changes-in-terraform-0-10/). The original comment is below._

<hr>

Support for datastore clusters was merged in https://github.com/vmware/govmomi/pull/428 I guess nothing stops us on terraform side? Waiting on this one as its a God sent tool :-)

**Comment by hashibot** [comment]
_This comment was originally opened by @mixacha as https://github.com/hashicorp/terraform/issues/3721#issuecomment-188981060. It was migrated here as part of the [provider split](https://www.hashicorp.com/blog/upcoming-provider-changes-in-terraform-0-10/). The original comment is below._

<hr>

@chrislovecnm itd be great if someone looked into this one now that vmware fully supports it

**Comment by hashibot** [comment]
_This comment was originally opened by @chrislovecnm as https://github.com/hashicorp/terraform/issues/3721#issuecomment-191971759. It was migrated here as part of the [provider split](https://www.hashicorp.com/blog/upcoming-provider-changes-in-terraform-0-10/). The original comment is below._

<hr>

@mixacha unfortunately I no longer have my vsphere test bed ;(

**Comment by hashibot** [comment]
_This comment was originally opened by @mixacha as https://github.com/hashicorp/terraform/issues/3721#issuecomment-192123277. It was migrated here as part of the [provider split](https://www.hashicorp.com/blog/upcoming-provider-changes-in-terraform-0-10/). The original comment is below._

<hr>

@chrislovecnm unfortunately :(
Really the last piece of puzzle for to enable enterprise companies to use the tool.

**Comment by hashibot** [comment]
_This comment was originally opened by @derng as https://github.com/hashicorp/terraform/issues/3721#issuecomment-208762728. It was migrated here as part of the [provider split](https://www.hashicorp.com/blog/upcoming-provider-changes-in-terraform-0-10/). The original comment is below._

<hr>

Same here. I need to demonstrate how powerful Terraform is and I'm from one of the biggest enterprise companies :(

**Comment by hashibot** [comment]
_This comment was originally opened by @GSkouroupathis as https://github.com/hashicorp/terraform/issues/3721#issuecomment-208818745. It was migrated here as part of the [provider split](https://www.hashicorp.com/blog/upcoming-provider-changes-in-terraform-0-10/). The original comment is below._

<hr>

+1

**Comment by hashibot** [comment]
_This comment was originally opened by @mixacha as https://github.com/hashicorp/terraform/issues/3721#issuecomment-209144653. It was migrated here as part of the [provider split](https://www.hashicorp.com/blog/upcoming-provider-changes-in-terraform-0-10/). The original comment is below._

<hr>

I've got a "small" lab that I can run tests against on any ESX version (currently 5.5 and 6 are installed in it).

**Comment by hashibot** [comment]
_This comment was originally opened by @chrislovecnm as https://github.com/hashicorp/terraform/issues/3721#issuecomment-209207954. It was migrated here as part of the [provider split](https://www.hashicorp.com/blog/upcoming-provider-changes-in-terraform-0-10/). The original comment is below._

<hr>

@mixacha appreciated. You develop at all?

**Comment by hashibot** [comment]
_This comment was originally opened by @kc11381 as https://github.com/hashicorp/terraform/issues/3721#issuecomment-209242087. It was migrated here as part of the [provider split](https://www.hashicorp.com/blog/upcoming-provider-changes-in-terraform-0-10/). The original comment is below._

<hr>

I will probably look at it. I had customized this plugin for my use case and I too needed Datastore cluster support. At that time I used some hack (basically joining paths) to support it and that is working fine. 
Now that proper support is available from govmomi api's, we should look at it.

**Comment by hashibot** [comment]
_This comment was originally opened by @chrislovecnm as https://github.com/hashicorp/terraform/issues/3721#issuecomment-209764950. It was migrated here as part of the [provider split](https://www.hashicorp.com/blog/upcoming-provider-changes-in-terraform-0-10/). The original comment is below._

<hr>

@kc11381 that would be aaaawwwweeesssoommmme!!!!

**Comment by hashibot** [comment]
_This comment was originally opened by @Shruti29 as https://github.com/hashicorp/terraform/issues/3721#issuecomment-211803384. It was migrated here as part of the [provider split](https://www.hashicorp.com/blog/upcoming-provider-changes-in-terraform-0-10/). The original comment is below._

<hr>

+1 badly need this :)

**Comment by hashibot** [comment]
_This comment was originally opened by @chrislovecnm as https://github.com/hashicorp/terraform/issues/3721#issuecomment-212036695. It was migrated here as part of the [provider split](https://www.hashicorp.com/blog/upcoming-provider-changes-in-terraform-0-10/). The original comment is below._

<hr>

@stack72 @phinze @jen20 
1. please remove the bug tag
2. please add feature

Everyone on this thread I will help this become a P1 feature.

**Comment by hashibot** [comment]
_This comment was originally opened by @chrislovecnm as https://github.com/hashicorp/terraform/issues/3721#issuecomment-212039311. It was migrated here as part of the [provider split](https://www.hashicorp.com/blog/upcoming-provider-changes-in-terraform-0-10/). The original comment is below._

<hr>

Pulling in couple more active devs and users:

@dkalleg @kristinn need some design ideas here.  Govomi has added support for datastores, https://github.com/vmware/govmomi/pull/428,  and now it needs to be implemented in TF.  Would either of you be willing to take this on?  I can but it may be a couple of weeks, and the scope is probably not huge.

@hugoboos @tommymalmqvist  @xantheran as vsphere gurus can you give us silly developers use cases, off of the top of my head:
1. what happens without DRS
2. what happens in a cluster
3. what happens without a cluser
4. what happenes without datastores
5. how does TF know which use case

@phinze the natives are saying that this has about as much need as windows support 😄

**Comment by hashibot** [comment]
_This comment was originally opened by @kristinn as https://github.com/hashicorp/terraform/issues/3721#issuecomment-212110160. It was migrated here as part of the [provider split](https://www.hashicorp.com/blog/upcoming-provider-changes-in-terraform-0-10/). The original comment is below._

<hr>

@chrislovecnm I would love to. Unfortunately I don't have access to a vmware instance with a datastore cluster configured. I've just used our small vmware setup at work so far while I'm creating PR for TF. 😞

**Comment by hashibot** [comment]
_This comment was originally opened by @chrislovecnm as https://github.com/hashicorp/terraform/issues/3721#issuecomment-212130232. It was migrated here as part of the [provider split](https://www.hashicorp.com/blog/upcoming-provider-changes-in-terraform-0-10/). The original comment is below._

<hr>

@kristinn any comments about design would be awesome.  Yah lots of bells and whistles ..

**Comment by hashibot** [comment]
_This comment was originally opened by @mixacha as https://github.com/hashicorp/terraform/issues/3721#issuecomment-214506359. It was migrated here as part of the [provider split](https://www.hashicorp.com/blog/upcoming-provider-changes-in-terraform-0-10/). The original comment is below._

<hr>

@kristinn  You can always run a hypervisor within a hypervisor + vCenter in trial mode for 60 days with highest license tier available and attach two VMDKs to form a datastore custer for testing purposes :-)

@chrislovecnm  I'm not a developer but logic inclines that terraform plugin should be able to determine regular datastore vs DS cluster (I believe these are marked as StoragePod type), and should a DS cluster be detected for placement, collect recommendations about initial placement - same as PowerShell cmd-lets are doing this. (also I think govc implemented a similar fashion to detect DS clusters and place virtual machines on top of them).

**Comment by hashibot** [comment]
_This comment was originally opened by @chrislovecnm as https://github.com/hashicorp/terraform/issues/3721#issuecomment-220424778. It was migrated here as part of the [provider split](https://www.hashicorp.com/blog/upcoming-provider-changes-in-terraform-0-10/). The original comment is below._

<hr>

This is a long time coming :) It will take 2+ PR's to get in. 

The first one https://github.com/hashicorp/terraform/pull/6777

**Comment by hashibot** [comment]
_This comment was originally opened by @chrislovecnm as https://github.com/hashicorp/terraform/issues/3721#issuecomment-223374936. It was migrated here as part of the [provider split](https://www.hashicorp.com/blog/upcoming-provider-changes-in-terraform-0-10/). The original comment is below._

<hr>

Still working on this ... But it is getting closer 😄

**Comment by hashibot** [comment]
_This comment was originally opened by @chrislovecnm as https://github.com/hashicorp/terraform/issues/3721#issuecomment-225674645. It was migrated here as part of the [provider split](https://www.hashicorp.com/blog/upcoming-provider-changes-in-terraform-0-10/). The original comment is below._

<hr>

If anyone on here is able to help testing, please let me know!!!  Code is here: https://github.com/hashicorp/terraform/pull/7031

Initial support for VMs.  Disks, Files, and CDs are not yet supported.

**Comment by hashibot** [comment]
_This comment was originally opened by @pfrayer as https://github.com/hashicorp/terraform/issues/3721#issuecomment-272415486. It was migrated here as part of the [provider split](https://www.hashicorp.com/blog/upcoming-provider-changes-in-terraform-0-10/). The original comment is below._

<hr>

Hello, any news on this feature, aka using datastore cluster name ?

Having a combo terraform + vmware for huge company is a must have. If stuff like datastore cluster doesn't work, is totaly unsable/useless. And then terraform just become a "tool used for POC or dirty testing", and not for an industrialized-enterprise-grade usage.

**Comment by hashibot** [comment]
_This comment was originally opened by @gbolo as https://github.com/hashicorp/terraform/issues/3721#issuecomment-284076510. It was migrated here as part of the [provider split](https://www.hashicorp.com/blog/upcoming-provider-changes-in-terraform-0-10/). The original comment is below._

<hr>

+1 

Would be nice to be able to deploy to a datastore cluster.

**Comment by hashibot** [comment]
_This comment was originally opened by @silveraid as https://github.com/hashicorp/terraform/issues/3721#issuecomment-284076868. It was migrated here as part of the [provider split](https://www.hashicorp.com/blog/upcoming-provider-changes-in-terraform-0-10/). The original comment is below._

<hr>

This would be really nice to have

**Comment by hashibot** [comment]
_This comment was originally opened by @renaudhager as https://github.com/hashicorp/terraform/issues/3721#issuecomment-288687372. It was migrated here as part of the [provider split](https://www.hashicorp.com/blog/upcoming-provider-changes-in-terraform-0-10/). The original comment is below._

<hr>

This feature is definitely needed.

**Comment by hashibot** [comment]
_This comment was originally opened by @Defilan as https://github.com/hashicorp/terraform/issues/3721#issuecomment-292016739. It was migrated here as part of the [provider split](https://www.hashicorp.com/blog/upcoming-provider-changes-in-terraform-0-10/). The original comment is below._

<hr>

Our environment relies heavily on datastore clusters and we use Terraform daily. This feature would be amazing!

**Comment by hashibot** [comment]
_This comment was originally opened by @paultravers as https://github.com/hashicorp/terraform/issues/3721#issuecomment-304074864. It was migrated here as part of the [provider split](https://www.hashicorp.com/blog/upcoming-provider-changes-in-terraform-0-10/). The original comment is below._

<hr>

This would be a killer feature. Currently trying to use a datastore cluster crashes vsphere.

**Comment by luntzel** [comment]
+1

**Comment by jaredmcook** [comment]
Is this issue still being addressed?  This is a really big deal for larger VMware shops.

**Comment by vancluever** [comment]
Hey @jaredmcook, this is definitely on the roadmap! It's not slated for our v1.0.0 release which will be in November, but it will be shortly thereafter - probably one of the first on-deck items after release. Stay tuned!

**Comment by tianomonteiro** [comment]
Hi @vancluever , we have any update about this? Thank you.

**Comment by pfrayer** [comment]
Is this issue still true with the new version of vSphere provider ?
https://www.hashicorp.com/blog/a-re-introduction-to-the-terraform-vsphere-provider

I don't have any vSphere around me to test it

**Comment by vancluever** [comment]
Hey all,

Storage cluster support is a feature that is still currently on the roadmap. Conservatively, I would expect it in a couple of months, released around the time of our support for DRS and storage DRS, but the actual release time may be different.

Hope this helps!

**Comment by pfrayer** [comment]
Thanks for the update @vancluever :) And it's nice to ear about DRS support !

**Comment by tianomonteiro** [comment]
Thanks for the update @vancluever .

**Comment by ranjanm1** [comment]
Hi, just checking is the feature has been progressed?

**Comment by jerrykan** [comment]
+1

**Comment by darkyat** [comment]
+1

DRS support would be really important. Right now I don't see any other way than manual datastore assignment when using local storage?

**Comment by c4nc** [comment]
+1

**Comment by NickLarsenNZ** [comment]
Is there something particularly difficult with Storage Clusters? Aren't they interacted with in the same way?

**Comment by vancluever** [comment]
Hi everyone,

I apologize for the delays here. As possibly mentioned in other threads, this is a feature we plan on working on very soon.

@NickLarsenNZ storage pods are not specifically datastores, and are actually more akin to VM folders in the vSphere API. So we need to go over the specifics of addressing them, in addition to evaluating the implications that working with them will have to the way we currently work with datastore attributes.

I understand that there is a lot of demand for this feature and know that we are for sure taking these requests seriously. However, as feedback has just recently been largely limited to +1's, update inquiries, and questions regarding the technical difficulty of implementing the feature, I am going to lock the thread for now until we have further developments on the issue.

Thanks!

**Comment by vancluever** [comment]
Hey all, I'm happy to say that we have released 1.4.0 with support for datastore clusters and Storage DRS. Check the [CHANGELOG](https://github.com/terraform-providers/terraform-provider-vsphere/blob/v1.4.0/CHANGELOG.md) for more details, in addition to the provider documentation for the following resources:

* [`vsphere_virtual_machine`](https://www.terraform.io/docs/providers/vsphere/r/virtual_machine.html)
* [`vshpere_datstore_cluster`](https://www.terraform.io/docs/providers/vsphere/r/datastore_cluster.html) (and the [data source](https://www.terraform.io/docs/providers/vsphere/d/datastore_cluster.html))
* [`vsphere_storage_drs_vm_override`](https://www.terraform.io/docs/providers/vsphere/r/storage_drs_vm_override.html)
* [`vsphere_nas_datastore`](https://www.terraform.io/docs/providers/vsphere/r/nas_datastore.html) and [`vsphere_vmfs_datastore`](https://www.terraform.io/docs/providers/vsphere/r/vmfs_datastore.html), which now support moving a datastore into a cluster

Thanks for all your patience on this one!

**Comment by Midacts** [comment]
Awesome. Thanks for adding this.
This is a big deal for vShpere VM deployments.
